### PR TITLE
Add vertex to cells map.

### PR DESCRIPTION
**Pull request by Rombur**
Return the active cells around each vertex and hanging node.

### CODE REVIEW

File: include/deal.II/grid/grid_tools.h [hunk 1]
```diff
@@ -630,6 +631,15 @@ namespace GridTools
   compute_ghost_cell_halo_layer (const Container &container);
 
 
+  /**
+   * Return the adjacent cells of all the vertices. The vertices are ordered by
+   * their ID.
```
Comment (bangerth) [inline include/deal.II/grid/grid_tools.h]: Can you rephrase the second sentence by using the term `index` instead of `ID` and then saying that `the vertex index is the number returned by calling <code>cell-@>vertex_index()</code>`? It may also be worthwhile mentioning that not all vertex indices are actually used; the used ones are marked in the array returned by `Triangulation::get_used_vertices()`.

File: source/grid/grid_tools.cc [hunk 2]
```diff
@@ -1595,6 +1595,37 @@ namespace GridTools
...
+    for (; cell!=endc; ++cell)
+      for (unsigned int i=0; i<GeometryInfo<dim>::vertices_per_cell; ++i)
+        vertex_to_cell_map[cell->vertex_index(i)].insert(cell);
+
+    // Take care of hanging nodes
+    cell = triangulation.begin_active();
+    for (; cell!=endc; ++cell)
+      for (unsigned int i=0; i<GeometryInfo<dim>::faces_per_cell; ++i)
+        {
+          if ((cell->neighbor_level(i)>=0) && (cell->neighbor_level(i)<cell->level()))
```
Comment (bangerth) [inline source/grid/grid_tools.cc]: The first part of this condition is equal to `cell->at_boundary(i) == false`, which is easier to read.

File: source/grid/grid_tools.cc [hunk 3]
```diff
@@ -1595,6 +1595,37 @@ namespace GridTools
...
+            {
+              typename Triangulation<dim,spacedim>::active_cell_iterator adjacent_cell =
+                cell->neighbor(i);
+              for (unsigned int j=0; j<GeometryInfo<dim>::vertices_per_face; ++j)
+                vertex_to_cell_map[cell->face(i)->vertex_index(j)].insert(adjacent_cell);
```
Comment (bangerth) [inline source/grid/grid_tools.cc]: This part of the algorithm, i.e., that cells are also counted as adjacent to hanging nodes located along their faces, needs to be described in the documentation.

The algorithm is only half-correct, though -- what do you do with hanging nodes located on edges in 3d?

Comment (Rombur) [inline source/grid/grid_tools.cc]: I have changed the documentation but I don't understand what's the problem with hanging nodes located on edges in 3D.

Comment (bangerth) [inline source/grid/grid_tools.cc]: Think of a situation where four cells come together at one edge, and one of them is refined. Then you add the two neighbors of the cell as adjacent to the central vertex of the (refined) edge. But you forget to add the cell that is "across the edge" from the refined cell because it is not a (coarser) neighbor of the child cells.

**Comment by Rombur** [comment]
Alright this should be fixed. I didn't realize that there was another function in ```GridTools``` that was doing something similar so I stole the solution for 3D problem :)

**Comment by bangerth** [comment]
Yes, this looks good. .Can you add a separate test to check the 3d case? You should be able to just create a 2x2x1 mesh where you refine one of the cells and ensure that all 3 of the active coarse level cells are adjacent to either 9 or 13 vertices.

**Comment by Rombur** [comment]
I have added the 3D test.

**Comment by bangerth** [comment]
Great, thanks!
